feat(core): add support for tags with `(print-)affected(:*)` and `run-many`

fguitton opened the pull request:
⚠️ _Edit: The behaviour of this pull request has changed to address feedbacks. See https://github.com/nrwl/nx/pull/10085#issuecomment-1493061063 for more information_

## Current Behavior
When running the `(print-)affected(:*)` command you get all affected projects for a target.
With the `run-many` command filtering is only possible with `--projects`.
Both accept a `--all`, it being warned for in `affected`.

## Expected Behavior
We would expect to be able to select projects by `tags`.

## This PR
This could be seen as a first step only acting to select projects in the same way `--projects` does it for `run-many`. Tags are not applied in compound but rather as `OR` statement in comma separated fashion.

That is in the example below the command would select projects that either contain the tag `sdk` or the tag `app`, not necessarily projects containing both tags.
```
nx affected --tags=sdk,app --target=test
```

## Related Issue(s)
This relates to very long standing #1621, #2675 or #8292 and a Nx **v14** compatible treatment while being different from #8364 which attempts to support globs but does not address `run-many`.

Closes #1621
Closes #2675

fguitton commented:
Hello @vsavkin !
I know you had interest in looking at this over the past few years.
It is the second PR opened for this feature but it addressed things slightly differently.
I am happy to bring any changes required, any feedback/direction is most welcome
We have a lot of need for this in our company. 

Thank you for your help and valuable time 🙏🏼 !

gioragutt commented on packages/nx/src/utils/project-graph-utils.ts:
```diff
@@ -29,6 +29,14 @@ export function projectHasTargetAndConfiguration(
 }
 
+export function projectHasTag(project: ProjectGraphProjectNode, tag: string) {
+  return !!(
+    project.data &&
+    project.data.tags &&
+    project.data.tags.includes(tag)
+  );
```
```suggestion
  return !!project?.data?.tags?.includes(tag);
```

fguitton replied:
Happy to bring the change ! This was done only to make the code style the same as for `projectHasTarget` and `projectHasTargetAndConfiguration` in the same file 😅

gioragutt replied:
I don't think it's a problem if you'll even fix the existing code, it's only for the better :)

fguitton replied:
That's now done in 9dafbd4, for other other methods too 😉

gioragutt commented:
Any update on this? I would love to use it 😁 Also, I thought it might be worth adding this to `print-affected` as well

fguitton commented:
> I thought it might be worth adding this to `print-affected` as well

Hello ! The `print-affected` command should already work out of the box with the changes proposed at it is considered a special branch of the `affected` command and the project filtering gets done upstream of that.

https://github.com/nrwl/nx/blob/7f5137da84e5d0f1b31f9011c57fc0e3bfaa9252/packages/nx/src/command-line/affected.ts#L37

https://github.com/nrwl/nx/blob/7f5137da84e5d0f1b31f9011c57fc0e3bfaa9252/packages/nx/src/command-line/affected.ts#L79-L113

I can imagine there is latent question with regard to the use/semantic of `--exclude` which only applies to project names. I do not necessarily have a set opinion on this. In effect a tag filtering can very much _replace_ the use of `--exclude`.

yharaskrik commented:
Would love to see this merged into Nx and my PR closed.

epechuzal commented:
This would be a big QoL feature for us, would love to see this in the next release :D

sylvainar commented:
Can't wait for this to be merged :smile:

josh-stevens commented:
pls

starch commented:
This would be a great QoL feature for our team 👍

jon9090 commented:
When this PR will be merged?

jon9090 commented:
@gioragutt

gioragutt commented:
> @gioragutt 

Can you add some context to why you tagged me? 😅

ianldgs commented:
Would really like this!

Use-case: only publish to the public npm registry packages tagged with `open-source`.

mrfelton commented:
Also interested in this. I want to be able to get a list of affected projects filtered by tag. Use case is for use with the nx dotnet  plugin. When I generate a dotnet app it actually generates 2 app. One is the main dotnet app. The other is a test app. I want to get a list of affected apps excluding the test apps.

msz13 commented:
When this feature will be published? I'm wating for it. Use case: create separate ci/cd pipelines for projects in different technologies, like js and dotnet (with nx dotnet plugin)

gioragutt commented:
@fguitton thank you for keeping this up to date for when they finally merge this 🙏🏻 I'm amused every time I get a notification about a rebase 😂

fguitton commented:
> I'm amused every time I get a notification about a rebase

Haha ! You're most welcome. I guess I'm being a bit overzealous on this one 😅 .
It's almost going to be the oldest non-merged (or closed) pull-request now ! 😄

mrfelton commented:
Is there anything specific that is blocking this from being merged?

fguitton commented:
@gioragutt 
> for when they finally merge this

We are now officially the older open pull request at sea. Perhaps this means we're next 😃

gioragutt commented:
> @gioragutt 
> 
> > for when they finally merge this
> 
> 
> 
> We are now officially the older open pull request at sea. Hehe. Perhaps this means we're next ! 😃 

That's actually funny 😄 and again, kudos to you for keeping it up all this time 🙏🏻

@

AgentEnder commented on docs/generated/packages/nx.json:
```diff
@@ -0,0 +1,345 @@
+{
```
I'm not sure this file should have been added. Can you `rm -rf docs && yarn documentation` on your next rebase?

AgentEnder reviewed (changes requested):
Overall this implementation looks good to me. I'll discuss more with @FrozenPandaz later, but on your next rebase can you double check the docs changes?

Also, I don't see any e2e tests added for this. If you could add a few (1 for affected, 1 for run-many) it would be really helpful.

I'm sorry this PR has kind of got left behind for a while, we will try to do better in the future for this.

AgentEnder commented on packages/nx/src/command-line/affected.ts:
```diff
@@ -162,6 +166,39 @@ function projectsToRun(
   return Object.values(affectedGraph.nodes) as ProjectGraphProjectNode[];
 }
 
+function affectedProjectsWithTags(
+  affectedGraph: ProjectGraph,
+  nxArgs: NxArgs
+): ProjectGraph {
+  const allProjects = Object.values(affectedGraph.nodes);
+  const matchingProjects = [];
+
+  allProjects.forEach((projectNode) => {
+    if (
+      findMatchingProjects(
+        nxArgs.tags,
+        projectNode.data.tags,
+        new Set(projectNode.data.tags)
+      ).length > 0
+    )
+      matchingProjects.push(projectNode.name);
+  });
+  const filteredProjects = matchingProjects.map(
+    (projectName) => affectedGraph.nodes[projectName]
+  );
+  const res = {
+    nodes: filteredProjects.reduce(
+      (nodes, project) => ({
+        ...nodes,
+        [project.name]: project,
+      }),
+      {}
+    ),
+    dependencies: affectedGraph.dependencies,
+  } as ProjectGraph;
+  return res;
+}
+
```
```suggestion
function affectedProjectsWithTags(
  affectedGraph: ProjectGraph,
  nxArgs: NxArgs
): ProjectGraph {
  const builder = new ProjectGraphBuilder(affectedGraph)
  const affectedProjectNodes = Object.entries(affectedGraph.nodes);

  for (const [nodeName, projectNode] of affectedProjectNodes) {
    if ( !hasMatchingTags(projectNode, nxArgs.tags ) {
      builder.removeNode(nodeName)
    }
  }
  
  return builder.getUpdatedProjectGraph();
}

```

AgentEnder reviewed (changes requested):
This can also be a bit simpler if we use the project graph builder.

NOTE - `hasMatchingTags` doesn't exist, you'll need to implement it. Most of the functionality should come from findMatchingProjects.

AgentEnder commented:
@fguitton Actually, after discussing this with @vsavkin a bit more today there is a bit more involved change that we'd like to make to properly support this and open some more doors in the future. Have you joined our community slack? I'd love to connect there and offer some more insight. 

Here's the link if you aren't there yet, just dm me 😄: https://go.nrwl.io/join-slack

fguitton commented:
@AgentEnder,
So I found a bit of time to explore the direction you indicated and I have updated the PR to use this new approach.

The `findMatchingProjects` function has been modified to a operated on the following pattern: `[tag:]filter[,...]`.
I have also removed the `Set(...)` performance bailout but making sure the performance test on CI were satisfied.

The principle stays the same and the change still covers `(print-)affected(:*)` and `run-many` but everything now happens through `--projects` argument. I have also made sure to permit a similar logic for the `--exclude` commutator as well, making sure is it backward compatible.

Negative filtering is supported and the syntax is flexible; either `!tag:ui` or `tag:!ui` indifferently.

Mix of project name and tag filters are also supported with an _non-chained_, negation-takes-all_ behaviour. Meaning the order of filters have no impact and we prefer to exclude in all cases.

I took the liberty to remove the mentions that tags only served for linting as well as added a couple of unit tests to insure that things behaved properly.

Feedback is obviously welcome, looking forward to this being merged in !

Best wishes 🌴

AgentEnder commented:
Awesome! Thanks for the updates @fguitton. I'll look over this either today or tomorrow.

AgentEnder commented on packages/nx/src/utils/find-matching-projects.ts:
```diff
@@ -1,46 +1,102 @@
+        if (minimatch.match(tags, pattern.value).length)
+          (pattern.exclude ? excludedProjects : selectedProjects).add(
+            projectName
+          );
```
Yeah, I'll look at this in the future

fguitton commented:
Thank you @AgentEnder ! Extending to the whole Nx team of course ! Thrilled to see this merged 🎉

gioragutt commented:
Congrats @fguitton, you've put a lot of work on this and it shows! Thank you for all your effort, I'm sure it'll bring immense value to a lot of people 🙏🏻🙏🏻🙏🏻
